### PR TITLE
Add caller attribution to tool execution structured logs

### PR DESCRIPTION
**Pull request by natel926**
### Description
When different agents or applications share a single MCP server instance, there is no way to attribute tool execution metrics to a specific caller. This adds support for an optional X-MCP-Caller HTTP header that agents or applications can set to identify themselves.

Changes:
- New caller_context.py: pure-ASGI middleware extracts X-MCP-Caller header into a contextvars.ContextVar (defaults to "unknown" if absent)
- tool_executor.py: includes "caller" field in every tool_execution log event
- streaming_server.py: registers the middleware on the Starlette app
- Tests for the middleware and the updated log field

The header is optional and backward-compatible — existing clients that do not send the header will appear as caller="unknown" in logs.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/mcp_server_opensearch/caller_context.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        """Extract caller header and forward to the wrapped app."""
+        if scope['type'] not in ('http', 'websocket'):
+            await self.app(scope, receive, send)
+            return
+
+        # Extract X-MCP-Caller from raw ASGI headers (list of [name, value] byte pairs).
+        caller = 'unknown'
+        for header_name, header_value in scope.get('headers', []):
+            if header_name == _HEADER_NAME:
+                caller = header_value.decode('latin-1')
```
Comment (mingshl) [inline src/mcp_server_opensearch/caller_context.py]: to avoid the malicious header, better to set a length limit to make sure the header_value is certain length limit. This header is also going to add to logs better to sanitize the length

**Comment by mingshl** [comment]
would you add some unit test?

**Comment by natel926** [comment]
> would you add some unit test?

Unit tests were already included (test_caller_context.py + test_tool_executor.py), but I've expanded them with 8 additional tests covering the sanitization edge cases (injection, oversized headers, etc.).

File: src/mcp_server_opensearch/caller_context.py (hunk 1 above)
Comment (natel926) [reply src/mcp_server_opensearch/caller_context.py]: Yes. we should set the length limit. Added _sanitize_caller() with a 64-char length limit and alphanumeric+hyphen+underscore+dot allowlist. Invalid values fall back to "unknown".

**Review by jiapingzeng — APPROVED** [review]

**Review by mingshl — APPROVED** [review]

**Comment by nathaliellenaa** [comment]
Thanks @natel926 , overall looks good! A couple of things:

- Docs: Could you add a short section to the user guide covering the X-MCP-Caller usage?
- SSE behavior: Over SSE, tools run in the long-lived `GET /sse` task, but tool calls come in as separate `POST /messages/` requests. So X-MCP-Caller only works if set on the `/sse` connection — set it on a `POST /messages/` request and it's ignored, falling back to unknown. Is per-message attribution the intent? If not, you can document it on the user guide as well

**Comment by natel926** [comment]
> Thanks @natel926 , overall looks good! A couple of things:
> 
> * Docs: Could you add a short section to the user guide covering the X-MCP-Caller usage?
> * SSE behavior: Over SSE, tools run in the long-lived `GET /sse` task, but tool calls come in as separate `POST /messages/` requests. So X-MCP-Caller only works if set on the `/sse` connection — set it on a `POST /messages/` request and it's ignored, falling back to unknown. Is per-message attribution the intent? If not, you can document it on the user guide as well

Good call — I'll add a section to USER_GUIDE.md covering X-MCP-Caller usage.
 
The CallerIdentificationMiddleware runs on all HTTP requests regardless of path or method, so it reads X-MCP-Caller from both GET /sse and POST /messages/. Clients that set the header at the HTTP client level will include it on every request automatically — no per-message setup needed. I'll add a note in the docs clarifying this. Will push an update shortly.

**Comment by Hailong-am** [comment]
> X-MCP-Caller

is it actual user of calling the tool?  From the PR description, it sounds like more than a identifier for client. how about named it as X-MCP-ClientId?

**Comment by natel926** [comment]
> > X-MCP-Caller
> 
> is it actual user of calling the tool? From the PR description, it sounds like more than a identifier for client. how about named it as X-MCP-ClientId?

Yes. It identifies the calling app/agent (for log attribution), not the end user. I avoided X-MCP-ClientId since client_id is already an OAuth term in the MCP SDK and would read as an auth credential. X-MCP-Caller is intentionally neutral, but happy to rename to X-MCP-Client-Name or whatever you prefer — just let me know.

**Comment by Hailong-am** [comment]
> X-MCP-Client-Name

This make sense to me. @mingshl @jiapingzeng would like to get your feedback as well

**Review by nathaliellenaa — APPROVED** [review]
Thanks @natel926 , LGTM!
